Make `std::os::darwin` public

madsmtm opened the pull request:
I'm not sure of the reasoning behind them not being public before, but I think they should be, just like `std::os::ios` and `std::os::macos` are public.

Additionally, I've merged their source code, as it was otherwise just a bunch of unnecessary duplication.

Ultimately, I've done this PR to fix `./x build library --target=aarch64-apple-tvos,aarch64-apple-watchos,aarch64-apple-visionos`, as that currently fails because of dead code warnings.

Since you reviewed https://github.com/rust-lang/rust/pull/121419
r? davidtwco

Fixes https://github.com/rust-lang/rust/issues/121640.

@rustbot label O-tvos O-watchos O-visionos

EDIT: Link to FCP summary: https://github.com/rust-lang/rust/pull/123723#issuecomment-2287276599

davidtwco commented:
r? libs

simlay commented on library/std/src/os/mod.rs:
I think all these `#[cfg(target_os = "ios/watchos/etc")]` could be one `#[cfg(target_vendor = "apple")]` and it would be a lot more readable.

madsmtm replied:
Hmm, but we still want to export them as separate modules `std::os::tvos` vs. `std::os::macos`?

simlay replied:
Ah yes, you're right I missed that.

madsmtm commented:
r? libs-api

dtolnay reviewed (changes requested):
Thanks for the PR!

I am open to adding these modules, but not the deprecated contents being exposed by this PR. `std::os::macos::raw` has been deprecated since more than 8 years ago and I don't think we should be adding more of that API for new targets.

madsmtm commented on triagebot.toml:
```diff
@@ -227,7 +222,7 @@ trigger_files = [
 
 [autolabel."O-macos"]
 trigger_files = [
-    "library/std/src/os/macos"
+    "library/std/src/os/apple"
 ]
```
Creating a more appropriate label for Apple-specific stuff is tracked in https://github.com/rust-lang/rust/pull/124146.

madsmtm reviewed (commented):
Hmm, the issue is that most of the `std::os::[os]::raw` definitions are re-exported in `std::os::unix::raw`, I think it's only the `stat` struct that isn't.

I've cfg-guarded that in the last commit, though I disagree that it's a good idea; the reason we'd remove it is to simplify `std` maintenance, but this actually only complicates it ;)

@rustbot ready

dtolnay commented:
@rust-lang/libs-api:

The `std::os::macos` module, which is available only on `cfg(target_os = "macos")`, contains:

<p align="center"><img src="https://github.com/rust-lang/rust/assets/1940490/c9ddb582-e6b4-44e1-9a9d-0dceae3a68ed" width="400"><br><img src="https://github.com/rust-lang/rust/assets/1940490/4f42fefd-8730-49a3-9a58-9a4123dc8bbc" width="400"><br><img src="https://github.com/rust-lang/rust/assets/1940490/2d1825c8-617f-4179-bd14-cc55a22fc4c4" width="400"><br><img src="https://github.com/rust-lang/rust/assets/1940490/586457a7-b592-4a2c-a7d1-f7df6ea091f6" width="400"><br><img src="https://github.com/rust-lang/rust/assets/1940490/29d76e5e-fd64-498e-838c-6511c2bdd629" width="400"></p>

This PR proposed exposing the same contents also under `std::os::tvos` for target_os="tvos", `std::os::visionos` for target_os="visionos", and `std::os::watchos` for target_os="watchos".

I propose that we accept instantly stabilizing these additions, with the following modifications:

- Do not expose `std::os::{tvos, visionos, watchos}::raw` &mdash; those modules will be changed to `pub(crate)`. For `macos`, `raw` has been deprecated since more than 8 years ago and I think it's worth avoiding adding that deprecated API in yet another place. Note that the 9 type aliases will continue to be accessible as re-exports in [`std::os::unix::raw`](https://doc.rust-lang.org/std/os/unix/raw/index.html), where they have been stably accessible on tvos, visionos, and watchos since the original introduction of these targets. For future unix targets, we may wish to avoid re-exporting raw deprecated types for them in `std::os::unix::raw` in the first place.

- Do not include `as_raw_stat` in the `MetadataExt` trait of the new targets.

- Seal the `MetadataExt` trait of the new targets.

Please register a concern if you would prefer that any of these modifications not be applied.

@rfcbot fcp merge

BurntSushi commented:
Is it possible to get a review/recommendation from the @rust-lang/apple team? I see that there are some folks [from that team](https://github.com/BurntSushi/team.git/blob/e1d15f97432ea3191ff114608260a06e44a17854/teams/apple.toml#L8-L13) involved here, but I don't see any sort of official consensus from that team. Basically, I don't have a ton of insight on whether this is the "right" organization for the Apple ecosystem (although I have no specific alternative suggestion), and I'd like to hear from _domain experts_ on this.

dtolnay commented:
@rustbot ping macos

BurntSushi commented:
I think the were recently renamed to `apple`? I assumed `@rust-lang/apple` would work, but I guess not.

@rustbot ping apple

BurntSushi commented:
OK, when in doubt, use brute force:

cc @hkratz @inflation @shepmaster @thomcc @nvzqz @madsmtm

libs-api is seeking you're opinion about stabilizing Apple specific APIs.

dtolnay commented:
Looks like "rustbot ping apple" is broken until https://github.com/rust-lang/rust/pull/124146 gets merged.

madsmtm commented:
I'm fine with implementing the proposed changes if that's what the decision ends up being, though I still disagree that it's a good idea - reiterating, the reason we'd remove it is to simplify `std` maintenance, but this only complicates it (as we have to keep the support for macOS/iOS anyhow).

Concrete example of an issue is that `MetadataExt::st_qspare` was added back in the day only on macOS, even though it is also available on the other platforms. Merging the files allowed us to see this, splitting them up or adding more `cfg`s will just lead to similar inconsistencies in the future.

bors commented:
:umbrella: The latest upstream changes (presumably #124146) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

dtolnay commented:
@rustbot ping apple

dtolnay commented:
@rustbot ping apple

inflation commented:
There could be an umbrella namespace for these common stuffs, e.g. `apple` like `cfg(target_vendor="apple")`.

joshtriplett commented:
That'd be my preference as well, here. If we have tvOS-specific APIs, by all means let's put them in `std::os::tvos`, but to the extent they're *identical* to macOS APIs, could we put them in the same place?

Amanieu commented:
`std::os::apple` would make sense since it is a `target_family` in the same way that `std::os::unix` is.

madsmtm commented:
> There could be an umbrella namespace for these common stuffs, e.g. `apple` like `cfg(target_vendor="apple")`.

I think `std::os::apple` is a good idea, but I think it's somewhat separate from this proposal; even if we have `std::os::apple`, I think I'd still want `std::os::tvos`/`std::os::watchos`/`std::os::visionos` for consistency with `std::os::macos`/`std::os::ios` (unless we fully deprecate the latter).

Implementation-wise, it could just be `pub use apple as ios;`.

Amanieu commented:
I would rather avoid having to include a new submodule every time apple designs a new device and instead centralize everything under a single submodule. For backwards-compatibility, `std::os::macos` can be a re-export of `std::os::apple`, but I don't think we should add more.

workingjubilee commented:
Hello,

I was there for some of the discussion of this nature because I reviewed the tvos PR and a few others. It basically went exactly like this discussion. There is no point in having all these different APIs exposed when they introduce a somewhat-pointless risk for additional variance.

It may be prudent to use the kernel name... i.e. `mod darwin`... that AIUI they all share, because there is The Other MacOS from long ago. Apple is clearly willing to have epochal breaking changes that risk redefining what `mod apple` would mean. Rosetta 2 implies its own sequel, a Rosetta III: Translator's Revenge.

madsmtm commented:
I opened https://github.com/rust-lang/rust/pull/125935 to do the refactor and fixing the compilation errors separately, so that this PR can be focused on the public API. Will rebase once that lands.

> It may be prudent to use the kernel name... i.e. `mod darwin`

Well, the kernel name is technically XNU, it's the broader core OS that's named Darwin, but your point still stands. I think it could be a pretty good name to use, were it not for the fact that the macOS targets use it in their names as well (e.g. `aarch64-apple-darwin`), which might be confusing?

A bit of prior art:
- Swift has `import Darwin` for Apple-specific stuff, and `#if os(Darwin) ... #endif` for Apple-specific configs.
- C/Objective-C has `TargetConditionals.h` which defines `TARGET_OS_MAC` for this purpose (`TARGET_OS_OSX` is the one for macOS-specific code).
- Python's `platform.system()` returns `Darwin` on macOS and `iOS` on iOS, while `os.uname()` returns `Darwin` on both.

---

In an ideal world, I think I'd want:
1. `cfg(target_family = "darwin")` (or just `cfg(darwin)`).
2. The macOS targets to be named `*-apple-macos`.
3. `std::os::darwin`.

So if we can agree that plan is to move in this direction (i.e. at _some_ point slowly rename the targets, and at _some_ point add `target_family = "darwin"`), then I think `std::os::darwin` makes sense; if not, then we should probably go with `std::os::apple`.

workingjubilee commented:
> In an ideal world, I think I'd want:
> 1. `cfg(target_family = "darwin")` (or just `cfg(darwin)`).
> 2. The macOS targets to be named `*-apple-macos`.
> 3. `std::os::darwin`.
>
> So if we can agree that plan is to move in this direction (i.e. at _some_ point slowly rename the targets, and at _some_ point add `target_family = "darwin"`), then I think `std::os::darwin` makes sense; if not, then we should probably go with `std::os::apple`.

Yes, I was in particular thinking that [the `target_family` cfg proposed for the phase out of `target_vendor`](https://github.com/rust-lang/rust/issues/100343) should be named thus in accord with this, and that ideally we would get `aarch64-apple-macos` as a target tuple (since the `target_os` is `"macos"` anyways).

joshtriplett commented:
We discussed this in today's @rust-lang/libs-api meeting. Our consensus was that we don't want to provide a different module for every Apple OS, and that we should only provide distinct modules if they have distinct functions that don't exist on the other OSes. Please *just* provide the `std::os::apple` module with the common functionality across all Apple OSes.

`macos` should re-export `apple` and also provide the deprecated items; `apple` shouldn't have the deprecated items.

joshtriplett commented:
It looks like https://github.com/rust-lang/rust/pull/125935 performed the merge we asked for; thank you!

Is there anything else that needs to happen here?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10377659089/job/28732338483) [(plain)](https://github.com/rust-lang/rust/commit/5ec86d72897f88675b7272a6a33d81a45aeb4f7f/checks/28732338483/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.749 Building wheels for collected packages: reuse
#16 2.751   Building wheel for reuse (pyproject.toml): started
#16 3.000   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.001   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#16 3.002   Stored in directory: /tmp/pip-ephem-wheel-cache-y6oonlqo/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.005 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.405 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.405 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.5s
---
    Checking addr2line v0.22.0
error[E0433]: failed to resolve: use of undeclared crate or module `raw`
  --> std/src/os/darwin/fs.rs:85:79
   |
85 |         unsafe { &*(self.as_inner().as_inner() as *const libc::stat as *const raw::stat) }
   |
help: consider importing one of these modules
   |
   |
6  + use crate::os::darwin::raw;
   |
6  + use crate::os::macos::raw;

For more information about this error, try `rustc --explain E0433`.
error: could not compile `std` (lib) due to 1 previous error
Build completed unsuccessfully in 0:01:12
```

</details>

madsmtm committed:
Make `std::os::darwin` public

This includes `std::os::darwin::fs`, which is re-exported under
`std::os::macos::fs` and `std::os::ios::fs`.

`std::os::darwin::raw` is not exposed, which means that
`MetadataExt::as_raw_stat` isn't available on tvOS, visionOS and
watchOS.

madsmtm commented:
I have updated this PR to instead make `std::os::darwin::fs` public. Since the exposed API surface here isn't really new, I propose to just make this insta-stable [^1].

[^1]: I considered going through the normal feature gate lifecycle, and making a tracking issue and whatnot, but from what I read in [here](https://rustc-dev-guide.rust-lang.org/stability.html#unstable), it seems like re-exporting `std::os::darwin::fs::*` under `std::os::macos::fs::*` with the old stability attributes might still end up making it stably accessible anyhow.

## Summary

`std::os::darwin` is available on `cfg(target_vendor = "apple")` (macOS, iOS, tvOS, visionOS and watchOS), and contains functionality specific to those operating systems. It is named "Darwin" because that's the name of the underlying operating system that all of these are built upon. This follows the naming convention in Swift (`import Darwin`), so while it's a little bit unknown, it should be fairly familiar to developers for Apple platforms.

The API for `std::os::darwin` is:

<p align="center">
<img width="400" alt="darwin" src="https://github.com/user-attachments/assets/870ef908-f9d0-4a3c-bc78-e75b8982b28d">
<img width="400" alt="darwin_fs" src="https://github.com/user-attachments/assets/1be83849-cc70-4fcb-9ff7-cefc69a2d951">
<img width="400" alt="darwin_fs_MetadataExt" src="https://github.com/user-attachments/assets/0cd135b4-99c6-458d-aebe-16378e8379aa">
<img width="400" alt="darwin_fs_FileTimesExt" src="https://github.com/user-attachments/assets/0197d584-89c4-4aaf-8b2a-4892fd0b7b1b">
</p>

The API for `std::os::macos` and `std::os::ios` is now:

<p align="center">
<img width="400" alt="macos_raw" src="https://github.com/user-attachments/assets/4b3095f8-25b1-4a02-868e-4ea7b2f9366f">
<img width="400" alt="macos_fs" src="https://github.com/user-attachments/assets/9f5386ea-63c6-47cc-86ca-909c78f6d98c">
</p>

Note: The images here doesn't show it, but `MetadataExt::as_raw_stat` is only available on macOS and iOS.

In the future, we can consider deprecating `std::os::macos` and `std::os::ios`, but this PR does not propose that.

@rustbot ready

madsmtm commented:
Responding to https://github.com/rust-lang/rust/pull/123723#issuecomment-2094885394:

> * Do not expose `std::os::{tvos, visionos, watchos}::raw` — those modules will be changed to `pub(crate)`. For `macos`, `raw` has been deprecated since more than 8 years ago and I think it's worth avoiding adding that deprecated API in yet another place. Note that the 9 type aliases will continue to be accessible as re-exports in [`std::os::unix::raw`](https://doc.rust-lang.org/std/os/unix/raw/index.html), where they have been stably accessible on tvos, visionos, and watchos since the original introduction of these targets. For future unix targets, we may wish to avoid re-exporting raw deprecated types for them in `std::os::unix::raw` in the first place.

Only `std::os::darwin::fs` is exposed, `std::os::darwin::raw` is `pub(crate)` and instead exposed publicly in `std::os::unix::raw`, `std::os::macos::raw` and `std::os::ios::raw`.

> * Do not include `as_raw_stat` in the `MetadataExt` trait of the new targets. On macos this trait method has been deprecated since Rust 1.8.

This is cfg-gated to only be available on macOS and iOS.

> * Seal the `MetadataExt` trait of the new targets.

I'm not entirely sure of a good way to do this without macros? `trait MetadataExt: #[cfg(...)] Sealed` doesn't work.

madsmtm commented:
r? libs-api

workingjubilee commented:
FCP checkboxes are back in this now-hidden comment, in case you're wondering why you can't find them: https://github.com/rust-lang/rust/pull/123723#issuecomment-2094885414

Dylan-DPC commented:
@joshtriplett @m-ou-se @BurntSushi this is waiting on you to vote/raise concerns on the [fcp](https://github.com/rust-lang/rust/pull/123723#issuecomment-2094885414). Thanks

dtolnay reviewed (approved):
Thank you!

dtolnay commented:
@bors r+

bors commented:
:pushpin: Commit 858d6d136a7943d79cacc0c24816e80d2a6ad67a has been approved by `dtolnay`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=dtolnay 858d6d136a7943d79cacc0c24816e80d2a6ad67a -->
<!-- homu: {"type":"Approved","sha":"858d6d136a7943d79cacc0c24816e80d2a6ad67a","approver":"dtolnay","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 858d6d136a7943d79cacc0c24816e80d2a6ad67a with merge 42188c3ca8ed84210c8cab6b1e2d553925e3dd2f...
<!-- homu: {"type":"BuildStarted","head_sha":"858d6d136a7943d79cacc0c24816e80d2a6ad67a","merge_sha":"42188c3ca8ed84210c8cab6b1e2d553925e3dd2f"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11655174720/job/32452016444)
Approved by: dtolnay
Pushing 42188c3ca8ed84210c8cab6b1e2d553925e3dd2f to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"dtolnay","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11655174720/job/32452016444"},"merge_sha":"42188c3ca8ed84210c8cab6b1e2d553925e3dd2f"} -->

rust-timer commented:
Finished benchmarking commit (42188c3ca8ed84210c8cab6b1e2d553925e3dd2f): [comparison URL](https://perf.rust-lang.org/compare.html?start=b8c8287a229cd79604aa84c25e1235fc78cd5f2e&end=42188c3ca8ed84210c8cab6b1e2d553925e3dd2f&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=b8c8287a229cd79604aa84c25e1235fc78cd5f2e&end=42188c3ca8ed84210c8cab6b1e2d553925e3dd2f&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.2% | [-1.2%, -1.2%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=b8c8287a229cd79604aa84c25e1235fc78cd5f2e&end=42188c3ca8ed84210c8cab6b1e2d553925e3dd2f&stat=max-rss)
<details>
<summary>Results (primary 0.2%, secondary 4.1%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.1%  | [2.1%, 2.1%]   | 1     |
| Regressions ❌ <br /> (secondary)  | 4.1%  | [4.1%, 4.1%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-1.8%, -1.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.2%  | [-1.8%, 2.1%]  | 2     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=b8c8287a229cd79604aa84c25e1235fc78cd5f2e&end=42188c3ca8ed84210c8cab6b1e2d553925e3dd2f&stat=cycles:u)
<details>
<summary>Results (secondary -2.6%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.6% | [-2.6%, -2.6%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=b8c8287a229cd79604aa84c25e1235fc78cd5f2e&end=42188c3ca8ed84210c8cab6b1e2d553925e3dd2f&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.466s -> 779.973s (-0.06%)
**Artifact size**: 335.33 MiB -> 335.34 MiB (0.00%)
